src/PostgREST/Plan.hs

```diff
@@ -363,6 +364,21 @@ mutatePlan mutation qi ApiRequest{..} sCache readReq = mapLeft ApiRequestError $
     rootOrder = maybe [] snd $ find (\(x, _) -> null x) qsOrder
     combinedLogic = foldr addFilterToLogicForest logic qsFiltersRoot
     body = payRaw <$> iPayload -- the body is assumed to be json at this stage(ApiRequest validates)
+    typedColumnsOrError = mapStopOnLeft (resolveOrError (HM.lookup qi $ dbTables sCache)) $ S.toList iColumns
```

Comment:
This looks up the table for each column right? Maybe we should cache the table like so
 
```suggestion
    tbl = HM.lookup qi $ dbTables sCache
    typedColumnsOrError = mapStopOnLeft (resolveOrError tbl) $ S.toList iColumns
```

---

Checking the [load tests](https://github.com/PostgREST/postgrest/pull/2542/checks?check_run_id=9363756197)

param | main | head
-- | -- | --
throughput | 317.3776065823794 | 308.3469937995588

We're losing some perf on the PR. Perhaps the above will help.

Reply:
Great comment, glad to see this level of focus on performance. That said, I think no, it doesn't look up the table more than once, more on that below.

First let me just comment on the performance impact overall of this part of the PR: I think the impact will be limited since the operation just adds a small constant time lookup per field, which should be dominated by the time already spent per field. The stuff postgresql does like parsing the JSON, parsing the SQL, doing the actual work of writing to the table; all of these have complexity linearish to _n_, _n_ being the column count, and each component alone is likely much larger than the time for this op (a hash map lookup in RAM).

Back to the table lookup. I'd expect GHC to only evaluate the `HM.lookup qi $ dbTables sCache` node once while partially applying `resolveOrError`. I'm no Haskell guy so might just be flopping about like a fish out of water here, but take a look at the generated core (`ghc -fforce-recomp -O2 -ddump-simpl`):

```
typedColumnsOrError_a2geX
  = hpc<PostgREST.Plan,179>
    hpc<PostgREST.Plan,178>
    $ (hpc<PostgREST.Plan,175>
       mapStopOnLeft_r2g9q
         (let {
            ds20_d2gYE :: Maybe Table
            [LclId]
            ds20_d2gYE
              = hpc<PostgREST.Plan,173>
                $ (hpc<PostgREST.Plan,170>
                   HM.lookup
                     PostgREST.SchemaCache.Identifiers.$fEqQualifiedIdentifier
                     PostgREST.SchemaCache.Identifiers.$fHashableQualifiedIdentifier
                     (hpc<PostgREST.Plan,169> qi_a2geh))
                  (hpc<PostgREST.Plan,172>
                   dbTables (hpc<PostgREST.Plan,171> sCache1_a2geC)) } in
          hpc<PostgREST.Plan,174>
          \ (ds21_d2gYF :: FieldName) ->
            hpc<PostgREST.Plan,168>
            case ds20_d2gYE of {
              Nothing ->
                hpc<PostgREST.Plan,157>
                Data.Either.Left
                  (hpc<PostgREST.Plan,156> PostgREST.ApiRequest.Types.NotFound);
              Just table_a2gf2 ->
                hpc<PostgREST.Plan,167>
                hpc<PostgREST.Plan,160>
                case resolveTableField
                       (hpc<PostgREST.Plan,158> table_a2gf2)
                       (hpc<PostgREST.Plan,159> ds21_d2gYF)
                of {
                  Nothing ->
                    hpc<PostgREST.Plan,164>
                    $ (hpc<PostgREST.Plan,161>
                       \ (ds22_d2gYL :: ApiRequestError) ->
                         Data.Either.Left ds22_d2gYL)
                      (hpc<PostgREST.Plan,163>
                       PostgREST.ApiRequest.Types.ColumnNotFound
                         (hpc<PostgREST.Plan,162> ds21_d2gYF));
                  Just typedField_a2gf4 ->
                    hpc<PostgREST.Plan,166>
                    Data.Either.Right (hpc<PostgREST.Plan,165> typedField_a2gf4)
                }
            }))
      (hpc<PostgREST.Plan,177>
       S.toList (hpc<PostgREST.Plan,176> ds11_d2h02)) } in
```

So `typedColumnsOrError_a2geX` applies `mapStopOnLeft_r2g9q` on a lambda which captures `ds20_d2gYE` from a `let` statement. `ds20_d2gYE` is basically the same as your `tbl = ...`. The first time we evaluate `case ds20_d2gYE of ...` in the lambda, I believe the "thunk" for `ds20_d2gYE` on the heap [should be overwritten](https://downloads.haskell.org/~ghc/7.8.3/docs/html/users_guide/informal-semantics.html) by a "WHNF", and so not evaluated again on subsequent calls to the same lambda. (And if I'm wrong here, the proposed version shouldn't make any difference anyhow since the only difference is that the binding happens a little earlier but still outside of the lambda).

Reply:
Broke out `tbl` anyhow, doesn't hurt to be more explicit that we want single eval on that expression.